lib/github/merge/local.ex

```diff
@@ -0,0 +1,229 @@
+defmodule BorsNG.GitHub.Merge.Local do
+  @moduledoc """
+  Uses local Git to construct staging/trying branches instead of GitHub's
+  API.
+
+  Only used when local merges are enabled.
+  """
+
+  use GenServer
+
+  alias BorsNG.Database.Repo
+  alias BorsNG.Database.Batch
+  alias BorsNG.Database.Project
+  alias BorsNG.Database.LinkPatchBatch
+  alias BorsNG.GitHub
+  alias BorsNG.GitHub.Merge.Hooks
+  alias BorsNG.Worker.Batcher
+
+  def start_link do
+    GenServer.start_link(__MODULE__, :ok, name: __MODULE__)
+  end
+
+  def init(_) do
+    {:ok, nil}
+  end
+
+  @doc """
+  GitHub allows GH Apps to authenticate over HTTPS using a username of
+  'x-access-token' and a password of the installation access token, which
+  happily gives us a way to authenticate CLI git commands.
+  """
+  def cli_creds(repo_conn) do
+    {:raw, token} = GitHub.get_raw_token!(repo_conn)
+    "x-access-token:#{token}"
+  end
+
+  @spec merge_batch!(Batch.t(), list(LinkPatchBatch.t())) ::
+          %{commit: String.t(), tree: String.t()} | :conflict
+  def merge_batch!(batch, patch_links) do
+    GenServer.call(__MODULE__, {:merge_batch, batch, patch_links}, :infinity)
+  end
+
+  @spec squash_merge_batch!(Batch.t(), list(LinkPatchBatch.t()), map) ::
+          %{commit: String.t(), tree: String.t()}
+  def squash_merge_batch!(batch, patch_links, toml) do
+    GenServer.call(__MODULE__, {:squash_merge_batch, batch, patch_links, toml}, :infinity)
+  end
+
+  def handle_call(msg, _from, state) do
+    reply = do_handle_call(msg)
+    {:reply, reply, state}
+  end
+
+  defp do_handle_call({:merge_batch, batch, patch_links}) do
+    batch = batch |> Repo.preload([:project])
+    workdir = batch.project.name
+
+    repo_conn = Project.installation_connection(batch.project.repo_xref, Repo)
+
+    # We make sure to clone the repository for each batch, rather than
+    # trying to reuse it, because hooks may change behavior in ways that could
+    # persist between batches in ways that Git can't detect, like modifying 
+    # the .git/config.
+    File.rm_rf!(workdir)
+    init_project_repo!(batch.project, cli_creds(repo_conn))
+
+    git = fn args -> System.cmd("git", args, cd: workdir) end
+
+    git.(["fetch", "origin", batch.into_branch])
+    git.(["checkout", "origin/#{batch.into_branch}"])
+
+    Hooks.invoke_before_merge_hook!(workdir)
+
+    merge_status =
+      patch_links
+      |> Enum.reduce(:merged, fn link_patch_batch, status ->
+        case status do
+          :conflict ->
+            :conflict
+
+          _ ->
+            link_patch_batch = link_patch_batch |> Repo.preload([:patch])
+            patch = link_patch_batch.patch
+
+            git.(["fetch", "origin", patch.commit])
+
+            {_, exit_code} =
+              git.([
+                "merge",
+                patch.commit,
+                "-m",
+                "[ci skip][skip ci][skip netlify] -bors-staging-tmp-#{patch.pr_xref}"
+              ])
```

Comment:
note that the commit message here is completely irrelevant: Bors only cares about the *contents* of the `staging.tmp` branch, which it uses to construct the actual merge commit manually. so these commit messages will never get seen. we at least set ci skip to avoid cluttering CI with irrelevant runs on the `staging.tmp` branch.